sup/gpu_uuid.cpp

```diff
@@ -0,0 +1,336 @@
+    //      GPU-f1fd7811-e4d3-4d54-abb7-efc579fb1e28
+    // becomes
+    //      f1fd7811e4d34d54abb7efc579fb1e28
+    std::remove_if(str, str+n, [](char c){return !std::isxdigit(c);});
+
+    // Converts a single hex character, i.e. 0123456789abcdef, to int
+    // Assumes that input is a valid hex character.
+    auto hex_c2i = [](unsigned char c) -> unsigned char {
+        c = std::tolower(c);
+        return std::isalpha(c)? c-'a'+10: c-'0';
+    };
+
+    // Convert pairs of characters into single bytes.
+    for (int i=0; i<16; ++i) {
+        const char* s = str+2*i;
+        result.bytes[i] = (hex_c2i(s[0])<<4) + hex_c2i(s[1]);
+    }
+
+    return result;
+}
+
+// For CUDA 9 the only way to get gpu uuid is via NVML.
+// NVML can be used to query all GPU devices, not just the
+// devices that have been made visible to the calling process.
+// Hence, there are two steps to finding the uuid of visible devices:
+// 1. Query the environment variable CUDA_VISIBLE_DEVICES to
+//    determine which devices are locally visible, and to enumerate
+//    them correctly.
+// 2. Query NVML for the uuid of each visible device.
+std::vector<uuid> get_gpu_uuids() {
+    // Get number of devices.
+    int ngpus = 0;
+    auto cuda_status = cudaGetDeviceCount(&ngpus);
+    if (cuda_status!=cudaSuccess) throw make_runtime_error(cuda_status);
+
+    // Attempt to initialize nvml
+    auto nvml_status = nvmlInit();
+    if (nvml_status!=NVML_SUCCESS) throw make_runtime_error(nvml_status);
+    auto nvml_guard = on_scope_exit(nvmlShutdown);
+
+    // store the uuids
+    std::vector<uuid> uuids;
+
+    // find the number of available GPUs
+    unsigned count = -1;
+    nvml_status = nvmlDeviceGetCount(&count);
+    if (nvml_status!=NVML_SUCCESS) throw make_runtime_error(nvml_status);
+
+    // Indexes of GPUs available on this rank
+    std::vector<int> device_ids;
+
+    // Test if the environment variable CUDA_VISIBLE_DEVICES has been set.
+    const char* visible_device_env = std::getenv("CUDA_VISIBLE_DEVICES");
+    // If set, attempt to parse the device ids from it.
+    if (visible_device_env) {
+        // Parse the gpu ids from the environment variable
+        device_ids = parse_visible_devices(visible_device_env, count);
+        if ((unsigned)ngpus != device_ids.size()) {
+            // Mismatch between device count detected by cuda runtime
+            // and that set in environment variable.
+            throw std::runtime_error(
+                "Mismatch between the number of devices in CUDA_VISIBLE_DEVICES"
+                " and the number of devices detected by cudaGetDeviceCount.");
+        }
+    }
+    // Not set, so all devices must be available.
+    else {
+        device_ids.resize(count);
+        std::iota(device_ids.begin(), device_ids.end(), 0);
+    }
+
+    // For each device id, query NVML for the device's uuid.
+    for (int i: device_ids) {
+        char buffer[41];
```

Comment:
There appears to be a constant for the maximum size of the UUID string returned, `NVML_DEVICE_UUID_BUFFER_SIZE`, which probably should be used instead.

Reply:
DONE.